Communications link between Arduino and Raspberry Pi using serial communications

noelpaz opened the pull request:
@russmcinnis  -- Please review with @mrrosen  and @CSTaylorJr 

I have merged the changes and tested compilation and running on the Arduino while connected to Raspberry Pi.

I had to add in Arduino FastCRC library: https://github.com/FrankBoesing/FastCRC
I cloned the master at commit 9be7f7164e74838df675df9703847baf35797988. The release 1.0 dated 2014 will not work, but that is the only release available.

russmcinnis committed:
added serial.cpp and serial.h

russmcinnis committed:
update link.cpp with some of the serial parts

russmcinnis committed:
update with serial added to latest link.cpp

Commit:
Revert "update with serial added to latest link.cpp"

This reverts commit 29cb0fecc6994f5b8b3774d60ab40bc0b87d0fb9.

Noel Paz - reverting commit so we can start from clean slate

Commit:
Added ommited struct on line 10

noelpaz committed:
Added serial tool

russmcinnis committed:
serial communication with data structures implemented.
Next crc16, check for response and timout.
put the data structure definition in the proper place
put the data and command buffers that are shared in the
proper place.

noelpaz committed:
Initial files for ecu ui link

russmcinnis committed:
update serial with ccitt crc16 check

russmcinnis committed:
update serial.cpp to ignore error for a test with raspberry pi

russmcinnis committed:
update with sequence number update

russmcinnis committed:
started moving packet code over from serial to link module

russmcinnis committed:
added data and command structures to link.h

russmcinnis committed:
update

russmcinnis committed:
update alarms

russmcinnis committed:
update alarms

noelpaz committed:
Added files for testing

noelpaz committed:
Class for module stub

noelpaz committed:
More functions added

noelpaz committed:
TestReader modified solved issue with timing

russmcinnis committed:
update for using uint data types

russmcinnis committed:
update for uint data types

russmcinnis committed:
update for uint data types

russmcinnis committed:
set receive index back to 0 in case of watchdog timeout

russmcinnis committed:
add crc just set not check

russmcinnis committed:
update data typ

russmcinnis committed:
added a serial debug using Serial1 pins 18 xmit, 19 rcv on DUE

russmcinnis committed:
added simple checksum and setting crc field to the checksum value

russmcinnis committed:
update baud to 38400 which was needed for raspberry pi

russmcinnis committed:
update for checksum

russmcinnis committed:
update link.h

russmcinnis committed:
the crc in the sending data packet wasn't actually getting set with the
checksum.

noelpaz committed:
Test reader

noelpaz committed:
Return and write to back to serial

noelpaz committed:
Additional prints to comfirm

noelpaz committed:
Added CRC for Out Packets

russmcinnis committed:
updated serial with update to data structures in comms

russmcinnis committed:
put data structs in comms

russmcinnis committed:
update link.h with structs for comms

russmcinnis committed:
remove start_byte from serial protocol

russmcinnis committed:
removed start_byte

russmcinnis committed:
removed start_byte from command structure

russmcinnis committed:
added fastcrc to use ccitt crc16 checking to match what python is using

noelpaz committed:
Removed folders and files not needed

russmcinnis committed:
updated debug for serial.cpp, put SERIAL_DEBUG in serial.h

russmcinnis committed:
put #define SERIAL_DEBUG in serial.h

russmcinnis committed:
put some debug in link.cpp

russmcinnis committed:
change debug

russmcinnis committed:
added debug to lin.h

russmcinnis committed:
updated to handle watchdog timeout properly

russmcinnis committed:
updated to allow for debug in link.cpp

russmcinnis committed:
reduced number of variables and put comments for each one

russmcinnis committed:
reduced number of variables and used a call by reference variable
for serial_ui and serial_debug

russmcinnis committed:
used call by reference serial_debug

russmcinnis committed:
moved clear crc alarm bit to when next valid package received.

russmcinnis committed:
removed unused command structures

russmcinnis commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+    serial_debug.println(" ");
+#endif 
+       
+    bytesSent = serial_ui.write((byte *)&comm.public_data_packet, sizeof(comm.public_data_packet));
```
The Serial.write() function returns before any byte is sent unless the data to write is longer than the allocated buffer size. It will block for the excess bytes. We have at least 70 bytes in the data packet so we will need to increase this size in HardwareSerial.h so that we are not blocking.
#if !defined(SERIAL_TX_BUFFER_SIZE)
#if ((RAMEND - RAMSTART) < 1023)
#define SERIAL_TX_BUFFER_SIZE 16
#else
#define SERIAL_TX_BUFFER_SIZE 64
#endif
#endif

there are other ways to do this 
https://www.hobbytronics.co.uk/arduino-serial-buffer-size

russmcinnis commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,119 @@
 // Link Module
 //
-
+#include <Arduino.h>
```
For using data types like uint32_t instead of unsigned long to make the structure easier to review byte mapping this include was required with Arduino.

mrrosen replied:
Shouldn't include Arduino.h -- use stdint.h for `uint32_t`/similar data types

russmcinnis commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,119 @@
 //
-
+#include <Arduino.h>
+#include <FastCRC.h>
```
https://github.com/FrankBoesing/FastCRC
this is supposed to be much faster.

mrrosen replied:
I would prefer to avoid any extra external dependencies when possible. I did try running their benchmarking code and while it is ~40% faster than the builtin function, its not a big deal; over the course of a second, we spend 20ms doing the crc (2% of the time: 2.5us per bit with builtin crc, 800 bits per packet, 10 packets per second to be overly generous with the numbers from what I was able to do). Bringing that down to 1.2% isnt really a big deal right now -- but we can look at this solution again if we need to save cycles. For now, please use the `_crc_xmodem_update` from `util/crc16.h` so we can avoid the external dependency

russmcinnis replied:
I received the mega and tried using utils/ccr16.h. I created a function that works fine in a standalone program using the data packet structure but when using with the controller it had strange behavior and caused serial protocol timeouts. I looked at the code and the functions are inline assembler in crc16.h. I have an option to compile with either. For now I am testing with the fastcrc library. Maybe there is a protothread issue with inline assembler.

russmcinnis committed:
updated to send data packet with crc without changes by other modules

russmcinnis committed:
update to send data structure with crc that will not change by making a
copy of public structure.

russmcinnis committed:
update link.h for data structure that will be sent.

russmcinnis commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,119 @@
+
+  PT_WAIT_UNTIL(pt, serialHalSendData() != HAL_IN_PROGRESS);
+
+    comm.public_data_packet.sequence_count = sequence_count;
```
this has been updated so that the the public data structure is copied and a crc produced to the copy that is sent to serial.cpp to write. this will keep the protocol logic in link.cpp and keep the crc valid while other modules are constantly updating the public data packet.

russmcinnis commented on src/modules/link.h:
```diff
@@ -2,10 +2,73 @@
+#include <Arduino.h>
 #include <stdint.h>
 
+#define ALARM_ECU_POWER_LOSS              0x01
```
need to see that this is up to date with latest protocol document. This is up to date with serialpacketv0.26

russmcinnis commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+    serial_debug.println(" ");
+#endif 
+       
+    bytesSent = serial_ui.write((byte *)&update_crc_data_packet, sizeof(update_crc_data_packet));
```
update_crc_data_packet is the copy of the public_data_packet that has the crc calculated. The public data structure can get updated before the send at this point so use this copy.

russmcinnis commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+    serial_debug.println(" ");
+#endif 
+       
+    bytesSent = serial_ui.write((byte *)&update_crc_data_packet, sizeof(update_crc_data_packet));
```
Serial.write() is non-blocking and returns before any bytes are sent as long as the data to write is less than the buffer size 64 bytes. The transmit buffer size is 64 bytes and the data packet is currently 70 bytes so there will be a block of 6 bytes. This buffer size is set in HardwareSerial.h. It is possible to create a board profile with an increase buffer size. https://www.hobbytronics.co.uk/arduino-serial-buffer-size. 
#if !defined(SERIAL_TX_BUFFER_SIZE)
#if ((RAMEND - RAMSTART) < 1023)
#define SERIAL_TX_BUFFER_SIZE 16
#else
#define SERIAL_TX_BUFFER_SIZE 64
#endif
#endif
#if !defined(SERIAL_RX_BUFFER_SIZE)
#if ((RAMEND - RAMSTART) < 1023)
#define SERIAL_RX_BUFFER_SIZE 16
#else
#define SERIAL_RX_BUFFER_SIZE 64
#endif
#endif
#if (SERIAL_TX_BUFFER_SIZE>256)
typedef uint16_t tx_buffer_index_t;
#else
typedef uint8_t tx_buffer_index_t;
#endif
#if  (SERIAL_RX_BUFFER_SIZE>256)
typedef uint16_t rx_buffer_index_t;
#else
typedef uint8_t rx_buffer_index_t;
#endif

There is also an Arduino function https://www.arduino.cc/reference/tr/language/functions/communication/serial/availableforwrite/
to see how much space is left in the send buffer. Maybe we call that and keep checking to ensure no blocking.

mrrosen replied:
Instead of relying on Serial.write to be non-blocking, since the packet size might change (ie, get bigger), it would be better to transfer only what you can at a time and return `HAL_IN_PROGRESS` in case you werent able to send everything:
```
// Assume function argument bufferLength for the length of message being sent
// and global bytesSent remembering how many bytes have been sent
int bytesToWrite = min(serial_ui.availableForWrite(), bufferLength - bytesSent);
bytesSent += serial_ui.write(&(buffer[bytesSent]), bytesToWrite);
if (millis() - startTime > timeout) {
  // Do here any state clearing needed so next call to serialHalSendData can start fresh
  bytesSent = 0;
  return HAL_TIMEOUT;
} else if (bytesSent < bufferLength) {
  return HAL_IN_PROGRESS
}
```

russmcinnis replied:
I will make this update.

russmcinnis commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+  }
+  if (watchdog_active == true)
+  {
+    if ((millis() - watchdog_start_ms) > watchdog_max_ms)
```
this has been tested, watchdog timeouts are handled so that the next data packet will be sent after a timeout.

russmcinnis commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+HardwareSerial &serial_debug = Serial1;
+#endif
+
+HardwareSerial &serial_ui = Serial;
```
change the Serial to Serial1, Serial2, etc. We will be moving away from the DUE programming port to serial pins on the Mega so this will need to change. There is also a debug port which could use Serial, so the ide or other serial terminal could be used to debug without an ftdi cable.

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+#include "../pt/pt.h"
+#include "../hal/hal.h"
+#include "../hal/serial.h"
+#include "../modules/link.h"
```
HAL layer should be completely independent of higher layers; so you should not need to include link.h

russmcinnis replied:
serial is now independent of link.h. The data and command structures from link.cpp are just bytes arrays with specified lengths for serial.cpp.

mrrosen commented on src/modules/parameters.h:
```diff
@@ -3,9 +3,7 @@
 #define __PARAMETERS_MODULE_H__
 
 #include <stdbool.h>
-
```
Shouldn't be modifying this file at all

mrrosen commented on src/modules/sensors.cpp:
```diff
@@ -30,6 +30,11 @@ static PT_THREAD(sensorsPressureThreadMain(struct pt* pt))
   pressureSensorHalGetValue(&rawPressure);
   
```
shouldnt be modifying this file at all

mrrosen commented on tools/serialtool/serialutil:
```diff
@@ -0,0 +1,542 @@
+#!/usr/bin/python
```
grabserial is licensed as GPLv2 -- wont probably be best not to include this code as part of an official distribution unless we know how to properly include it

russmcinnis replied:
I will talk to Noel. I don't think he needs this here anymore.

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+  // return MODULE_FAIL;
+  do
+  { serial_ui.read();
+  } while (serial_ui.available() > 0);
```
Might be good to move this code into a local function (`serialClear()` maybe)

russmcinnis replied:
serialClear() has been added and called in two places.

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+  return HAL_OK;
+}
+
+int serialHalGetData(void)
```
Would be better to pass in pointer and length to a buffer stead of global variables to transfer deata to serialHalGetData; like storage HAL API -- would still need timeout mechanism

russmcinnis replied:
currently serial.cpp is using a bytearray from link.cpp. extern for address and size currently. So will re-visit this.

russmcinnis replied:
I could probably pass in pointer, currently validating all parameters to RPi so can get back to seeing if I can do this without global. Currently getting generic byte array global and size variable global for the send and receive buffers.

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+    command_packet_u.command_bytes[incoming_index] = inByte;
+    
+    incoming_index++;
+
```
index can be a local static instead of global

russmcinnis replied:
I had left it global instead of static because I reset it to 0 just after sending a data packet. The function is currently getting one byte at a time and returning. I can probably do this another way if needed.

russmcinnis replied:
I was able to make it static local as well as a few other places.

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+}
+
+int serialHalSendData()
+{
```
Pass in buffer, length, and timeout like `serialHalGetData` instead of globals; see above

russmcinnis replied:
currently using generic byte array and length globals but not using global indexes.

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+int serialHalSendData()
+{
+  current_send_ms = millis();
+  if ((current_send_ms - last_send_ms) >= send_interval_ms)
```
Periodic timer for sending a message should be handled in the link module

CSTaylorJr reviewed (commented):
Quick feedback:

Overall it looks good. Haven't tried running it. Did you change the SERIAL_RX_BUFFER_SIZE in HardwareSerial.h to 128 bytes? If so, we need to make sure that gets documented, because that file isn't in the repo.

Alternatively, I think we could shrink the data packet to 64 bytes or less. The pressure data is converted to Pascals, and ranges from +/-25,000, which easily fits into a 16 bit int. I suspect many of the other 32 bit values can be shrunk to 16 bits as well (rate, volume).

I'll give the code more scrutiny tonight.

russmcinnis commented:
@CSTaylorJr - I would like to follow the serial guide for the updates so if we change sizes we can use this as our guide. Or if we need the code ASAP let me know which fields change so I can communicate with Noel on the Raspberry Pi. The serial protocol is based off of Michael's last update serialpacket0.26.ppt from the software slack channel https://aqualungworkspace.slack.com/files/U01022R8YKU/F0118FRCDNZ/serialpacketv0.26.pptx

mrrosen commented on src/hal/serial.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+       
+    //last_sequence_count = sequence_count;
+    sequence_count++;
```
sequence_count should be managed by the link module

russmcinnis replied:
this has been resolved so there is no sequence count reference in serial.cpp

mrrosen commented on src/hal/serial.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef __SERIAL_HAL_H__
+#define __SERIAL_HAL_H__
+
+//#define SERIAL_DEBUG
```
Remove dead code

mrrosen commented on src/hal/serial.h:
```diff
@@ -0,0 +1,13 @@
+#include "../hal/hal.h"
+
+int serialHalInit(void);
+int serialHalGetData(void);
```
Please add //TODO: Doc on the tops of each of these -- we should add comments latter on each functions to define exactly what it does

russmcinnis replied:
updated serial.cpp, serial.h, link.cpp and link.h with comments

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
-// TODO: Fix this variable name :(
+// Private Variables
 struct link comm;
+static struct pt serialThread;
```
Please separate the private (Static) and public variables as the other modules do

russmcinnis replied:
variables in serial.cpp and link.cpp have been separated like other modules.

mrrosen commented on src/modules/link.h:
```diff
@@ -2,10 +2,73 @@
 #define __LINK_MODULE_H__
 
+#include <stdbool.h>
+#include <Arduino.h>
```
Dont need Arduino.h here

russmcinnis replied:
removed Arduino.h from link.h

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
 {
-  // TODO: Implement
-  return MODULE_OK;
+  if (serialHalInit() != HAL_OK) {
```
Need to PT_INIT any threads used

russmcinnis replied:
added PT_INIT for serial send and read threads

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
+  PT_BEGIN(pt);
+
+  if (!PT_SCHEDULE(serialReadThreadMain(&serialReadThread))) {
+    PT_EXIT(pt);
```
Since the protocol we defined is just `Start Timer (100ms) -> Form Packet -> Send Packet (serialHalSendData) -> Read Back Response (serialHalGetData(50ms timeout)) -> Check Response -> Update Public Block -> Wait on Timer (100ms)`, this can all be done in a single thread. Initially, when I proposed two threads, I thought we might have a more asynchronous protocol (RPi would send data at any time, ECU might send a new packet any time instead of this 10Hz cadence). The timer hal would make implementing this easy.

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
+  PT_BEGIN(pt);
+
+  PT_WAIT_UNTIL(pt, serialHalSendData() != HAL_IN_PROGRESS);
+    memcpy((void *)&update_crc_data_packet, (void *)&comm.public_data_packet, sizeof(update_crc_data_packet));
```
You need to collect the various fields for the data packet from the public interfaces on the other modules -- for example the current set points can be retrieved from the parameters module, the pressure measurements from the sensors module, etc

mrrosen commented on src/modules/link.h:
```diff
@@ -2,10 +2,73 @@
+// bits 28 - 31 --
+
+
+#define PACKET_VERSION 1
```
Packet version should be 0x02 for our current one

mrrosen commented on src/modules/link.h:
```diff
@@ -21,6 +84,14 @@ struct link {
 // Public Variables
 extern struct link comm;
 
+extern uint16_t sequence_count;
```
Shouldnt need any other public variables than those in the interface structure `struct link`

russmcinnis replied:
no other public variables in link.h other than struct link

mrrosen commented on src/modules/link.h:
```diff
@@ -2,10 +2,73 @@
 struct link {
-  // Variables
+  bool update;
+  bool assist;
```
these booleans are no longer needed and should be removed

russmcinnis replied:
removed these booleans

mrrosen commented on src/modules/link.h:
```diff
@@ -2,10 +2,73 @@
-  // Variables
+  bool update;
+  bool assist;
+  data_packet_def public_data_packet;
```
shouldnt just post the command and data packets in the public interface -- the module needs to parse of the needed fields from the command packet and set the `startVentilation`, `ventiliationMode`, etc fields in the `struct link`; all the data needed for a data packet should come from other modules interface structures, see comments above.

russmcinnis replied:
these structures have been moved out of the link.h public interface and copied to link.cpp where they are used to copy to and from variables from other modules.

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
+
+  if (watchdog_exceeded == true)
+  {
+    comm.public_command_packet.alarm_bits |= 1 << ALARM_DROPPED_PACKET;
```
`ALARM_DROPPED_PACKET` doesnt define an offset, but the bit itself (which is a style I do like better), so this should be `alarm_bits |= ALARM_DROPPED_PACKET` instead of `alarm_bits |= 1 << ALARM_DROPPED_PACKET`

russmcinnis replied:
I have fixed the action to match the bit pattern.

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
+      serial_debug.print("unexpected sequence count: ");
+      serial_debug.println(comm.public_command_packet.sequence_count, DEC);
+#endif    
+      comm.public_command_packet.alarm_bits |= 1 << ALARM_CRC_ERROR;
```
An incorrect sequence number is actually more an indication of a dropped packet than a bad packet, so should set DROPPED_PACKET instead of CRC_ERROR

russmcinnis replied:
changed it to dropped packet

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
+    }
+    else
+    {
+      calc_crc = CRC16.ccitt((uint8_t *)&comm.public_command_packet, sizeof(comm.public_command_packet) - 2);
```
As mentioned above, can be replaces with (note, pseudo code) 
```
uint16_t crc = 0;
for (int i = 0; i < sizeof(command_packet); i++) {
  crc = _crc_xmodem_update(crc, command_packet[i]); // Note, from the code Ive seen, Noel's python library uses xmodem crc, not ccitt -- just be sure to use the same crc algo and initializer on both sides
}
```
to avoid external dependency

russmcinnis replied:
I have a function that wraps the avr c function. it is causing issues. maybe the inline assembly straight from the header source..

russmcinnis replied:
I get many packets through but get many timeouts. I see 50 msec gaps in a couple of places. It might not be noticeable unless specifically running serial debug mode. I will leave fastcrc for the default until this is fixed with standard avr c. here is the function I use which is identical to the one I can test with a standalone application.  checking time it is 2 - 4 msec for either crc in the standalone app. Fastcrc is normally around 1 msec faster in the standalone app. We do not see the 50msec gaps in the standalone app with the avr C crc.

russmcinnis replied:
I am not sure why, but it is now running without the timeouts. I will check it in with switch set for avr C crc16. I will keep an eye on this in serial debug mode.

mrrosen commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,123 @@
+    memcpy((void *)&update_crc_data_packet, (void *)&comm.public_data_packet, sizeof(update_crc_data_packet));
+    update_crc_data_packet.sequence_count = sequence_count;
+    update_crc_data_packet.packet_version = PACKET_VERSION;
+    update_crc_data_packet.crc = CRC16.ccitt((uint8_t *)&comm.public_data_packet, sizeof(comm.public_data_packet) - 2);
```
see above about crc calculations

russmcinnis replied:
updated to use avr c library by default. rewrote the function and it is running without timeout issues that I was seeing before.

mrrosen reviewed (changes requested):
Here is my initial review -- quite a few notes but its on the right track

russmcinnis committed:
use last_sequence_count variable

russmcinnis committed:
use last_sequence_count variable

russmcinnis committed:
fixed debug messages

russmcinnis committed:
fixed debug messages

russmcinnis committed:
add packet total, dropped packet total and timeouts to debug.
also send proper data for the first packet.

russmcinnis committed:
stdint.h

russmcinnis committed:
stdint.h

russmcinnis committed:
use tmp storage for command packet in cases of bad data

russmcinnis committed:
use temporary storge for command data so it checks out
before copying to public. also do not write alarm bits.

russmcinnis committed:
updated sequence time to 100 msec

russmcinnis committed:
update to show byte count when timeout

russmcinnis committed:
update interval back to 100

russmcinnis committed:
remove excess debug

russmcinnis committed:
fixed watchdog process so the watchdog active flag is checked before
sending more data

russmcinnis committed:
turn off debug

russmcinnis committed:
updated debug messages

russmcinnis committed:
changed watchdog_active to read_active which made more sense. We
are now checking this flag before sending bytes. Also before
reading bytes. This was needed to sync up so we should not
get dropped packets or watchdog timeouts.

russmcinnis committed:
removed read_active variable that was not used in link module. only used
in serial.cpp

russmcinnis committed:
added debug to note that read queue is empty.

russmcinnis committed:
removed debug that was causing issues for Noel

russmcinnis committed:
put back earlier commit change

russmcinnis committed:
put back earlier commit change

russmcinnis committed:
put back earlier commit

russmcinnis committed:
turn off debug port flag

russmcinnis committed:
set and clear dropped packet and crc alarm bits

russmcinnis committed:
update serial.cpp for 70 msec watchdog

russmcinnis committed:
updated link.cpp with command packet interpreter

russmcinnis committed:
updated link.h with mode bits

russmcinnis committed:
update values from command packet

russmcinnis committed:
debug with data packet sending back set values and measured values

russmcinnis committed:
link.cpp - #ifdef some debug to validate RPi and Arduino communication

russmcinnis committed:
remove debugging serial.print() with a bunch of byte writes causing
timing issues

russmcinnis committed:
updated watchdog timeout to 200 msec

russmcinnis committed:
updated link.cpp for debug messages

russmcinnis committed:
updated serial.cpp to use macros for Serial, Serial1
also macros for settings (BAUD, interval, watchdog)
these are defined in serial.h

russmcinnis committed:
updated serial.h to use #define for serial ports and other
settings (interval, watchdog, baud)

russmcinnis committed:
updated debug to use macros.

russmcinnis committed:
removed data structure dependency from serial.cpp, use generic byte
array to send data packet

russmcinnis committed:
use macros for Serial

russmcinnis committed:
removed serial dependencies.

russmcinnis committed:
removed debug and comments not needed

russmcinnis committed:
removed debug and comments not needed

russmcinnis committed:
comment at top for document

russmcinnis committed:
comment at top for document

russmcinnis committed:
separate public and private variables like other modules

russmcinnis committed:
separate public and private variables like other modules

russmcinnis committed:
PT_INIT the serial send and read threads

russmcinnis committed:
added function that wraps utils/crc16.h function to get the
same value for crc16 that we get with fastcrc. There is a
a #define USE_DUE that is commented by default so this
function will be used for crc calculations with the mega.
tested the function using an uno.

russmcinnis committed:
updated funcion to set data packet fields from parameters, control and
sensor modules

russmcinnis committed:
changed inComingIndex to be static local since only used in one function
it can be reset to 0 within the function.

russmcinnis committed:
updated serial.cpp to use available bytes in the writes

russmcinnis committed:
serial.h updated with more #define setup

russmcinnis committed:
option #define USE_FAST_CRC since avr crc16 causing
issues

russmcinnis committed:
put in send timeout send sends are broken up into sizes that fit into
the current serial.write buffer size to keep from blocking. timeout
is set for 30msec. set to 1 for testing and it would often timeout and
proceed with sending next data packet. do not read anything after this
just set timeout and proceed,

russmcinnis committed:
changed public variable to private

russmcinnis committed:
turned off debug by default

russmcinnis committed:
in link.cpp to help ui validate some data fields

russmcinnis committed:
test values for UI in link.cpp #ifdef SET_VALUES

russmcinnis committed:
add tidal volume measured value from current volume sensor module to data packet

russmcinnis committed:
put comment in function that sets data packet fields to send back
that respiratory rate was not found. looked in control.h struct.

russmcinnis committed:
moved several private global variables into static declared in functions
where needed.

russmcinnis committed:
moved simulation #ifdef SET_VALUES to a place where they will not be
overwritten by the unset module variables for ui end to end testing

russmcinnis committed:
updated the watchdog timeout for arduino to 200. It was left at 1000msec
by accident.

dawnlstone commented on src/modules/link.cpp:
```diff
@@ -1,22 +1,285 @@
+  comm.public_data_packet.ie_ratio_set = comm.public_command_packet.ie_ratio_set;
+  comm.public_data_packet.tidal_volume_measured = 100;
+  comm.public_data_packet.respiratory_rate_measured = 200;
+  comm.public_data_packet.ie_ratio_set = 300; 
```
s/b "ie_ratio_measured = 300"

russmcinnis replied:
- [x] I have now removed the simulation mode line where ie_ratio_set is hardcoded to 300. It is now set to the command packet value. I probably did mean to set the measured value in the first place.

russmcinnis committed:
set ie rate test value to incoming command packet set value

russmcinnis committed:
moved public data and command structures out of the public interface in
link.h so updated link.cpp with the copies only in that module.

russmcinnis committed:
moved public data and command structures out of link.h

russmcinnis committed:
added serialClear() function since the operation was repeated

russmcinnis committed:
replace avr c function wrapper with the same text used in a standalone
test. The function gives a good crc, but the serial watchdog will get
many timeouts. You might not notice it unless you have SERIAL_DEBUG
turned on. Leaving default to fastcrc.

russmcinnis committed:
crc16 from avr lib c is now working without timeout issues

Commit:
serial: remove unused serialtool

Signed-off-by: Rosen, Michael R <michael.r.rosen@intel.com>

russmcinnis committed:
fix mode value logic

mrrosen reviewed (approved):
Rebased branch and tried it against UI; seems to work fine
